haystack/utils/type_serialization.py

```diff
@@ -20,54 +21,66 @@ def serialize_type(target: Any) -> str:
     Serializes a type or an instance to its string representation, including the module name.
 
     This function handles types, instances of types, and special typing objects.
-    It assumes that non-typing objects will have a '__name__' attribute and raises
-    an error if a type cannot be serialized.
+    It assumes that non-typing objects will have a '__name__' attribute.
 
     :param target:
         The object to serialize, can be an instance or a type.
     :return:
         The string representation of the type.
-    :raises ValueError:
-        If the type cannot be serialized.
     """
-    # If the target is a string and contains a dot, treat it as an already serialized type
-    if isinstance(target, str) and "." in target:
-        return target
-
-    # Determine if the target is a type or an instance of a typing object
-    is_type_or_typing = isinstance(target, type) or bool(get_origin(target)) or target == Any
-    type_obj = target if is_type_or_typing else type(target)
-    type_obj_repr = repr(type_obj)
-
-    if type_obj_repr.startswith("typing."):
-        # e.g., typing.List[int] -> List[int], we'll add the module below
-        type_name = type_obj_repr.split(".", 1)[1]
-    elif origin := get_origin(type_obj):  # get the origin (base type of the parameterized generic type)
-        # get the arguments of the generic type
-        args = get_args(type_obj)
-        args_repr = ", ".join(serialize_type(arg) for arg in args)
-        type_name = f"{origin.__name__}[{args_repr}]"
-    elif hasattr(type_obj, "__name__"):
-        type_name = type_obj.__name__
-    else:
-        # If type cannot be serialized, raise an error
-        raise ValueError(f"Could not serialize type: {type_obj_repr}")
+    name = getattr(target, "__name__", str(target))
+
+    # Remove the 'typing.' prefix when using python <3.9
+    if name.startswith("typing."):
+        name = name[7:]
+    # Remove the arguments from the name when using python <3.9
+    if "[" in name:
+        name = name.split("[")[0]
 
-    module = inspect.getmodule(type_obj)
+    # Get module name
+    module = inspect.getmodule(target)
     if module and hasattr(module, "__name__"):
         if module.__name__ == "builtins":
             # omit the module name for builtins, it just clutters the output
             # e.g. instead of 'builtins.str', we'll just return 'str'
-            full_path = type_name
+            module_name = ""
         else:
-            full_path = f"{module.__name__}.{type_name}"
+            module_name = f"{module.__name__}"
     else:
-        full_path = type_name
+        module_name = ""
```

Comment:
just an idea...

Could this part be simplified if we initially set `module_name=""` and then we modify it where appropriate?